### PR TITLE
user DQ preference settings DQ#222

### PR DESCRIPTION
**Pull request by alexhuang091**
DQ#222 user preference settings.

- This feature allows user to set his preferred profile and expand/collapse state of profile details.
- The preference settings is saved to userdetails if logged in or cookie (expiry = 365 days) if not logged in.
- Once user clicks 'save' on settings dialog, any selected data profile in current url will be removed and current profile being applied and webpage reloaded.
- There's also a requirement that if the user changed the expand/collapse state of the profile details, next time when page reload, the state will be kept. The solution is, every time user expands/collapses the detail, the state is written into a session cookie. Also when user saves the preference, the expand/collapse preference is written into the same cookie so next time page reload we can correctly set the expand/collapse state.
- Also the postion of tooltips for settings and download link are changed to above the control.

### CODE REVIEW

File: grails-app/assets/javascripts/search.js [hunk 1]
```diff
@@ -622,6 +622,142 @@ $(document).ready(function() {
...
+        if (BC_CONF.userId) {
+            // save the dq profile detail expand/collapse state
+            $.cookie(BC_CONF.expandKey, {expand: userPref.expand});
+            $.ajax({
+                url: BC_CONF.serverName + "/user/" + BC_CONF.prefKey,
+                type: "POST",
+                contentType: 'application/json',
+                data: JSON.stringify(userPref),
+                success: applyUserPreference // reload on success
+            }).always(function() {
```
Comment (nickdos) [inline grails-app/assets/javascripts/search.js:688]: Might need to add a `.fail()` or equivalent here to show user error message if this call fails.

File: grails-app/conf/plugin.groovy [hunk 2]
```diff
@@ -29,9 +29,11 @@
 auth.admin_role = "ROLE_ADMIN"
 serverName = 'http://dev.ala.org.au:8080'
 
-dataquality.enabled = false
-dataquality.baseUrl = 'https://biocache.ala.org.au/data-quality/'
+dataquality.enabled = true
```
Comment (nickdos) [inline grails-app/conf/plugin.groovy:32]: Not sure it should ON be default - most LA deployments will not use it.

File: grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy [hunk 3]
```diff
@@ -62,9 +63,10 @@ class OccurrenceController {
     def list(SpatialSearchRequestParams requestParams) {
         def start = System.currentTimeMillis()
 
-        def activeProfile = time("active profile") { qualityService.activeProfile(requestParams.qualityProfile) }
-
-        normaliseRequestParams(requestParams)
+        // get the user preference settings
+        def userPref = userService.getUserPref(authService?.getUserId(), request)
```
Comment (nickdos) [inline grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy:67]: This code is not surrounded by check if `dataquality.enabled == true`. Should it be? Checking if you have run it with that config set to false?

File: grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy [hunk 4]
```diff
@@ -160,16 +162,23 @@ class OccurrenceController {
             }
 
             def qualityCategories = time("quality categories") { qualityService.findAllEnabledCategories(requestParams.qualityProfile) }
-            def qualityFiltersByLabel = time("quality filters by label") { qualityService.getEnabledFiltersByLabel(requestParams.qualityProfile) }
+            def qualityFiltersByLabel = [:]
```
Comment (nickdos) [inline grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy:165]: Same here

File: grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy [hunk 5]
```diff
@@ -505,6 +505,10 @@ class PostProcessingService {
     }
 
     def processUserFQInteraction(SpatialSearchRequestParams requestParams, activeFacetObj) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy:507]: Service methods should have typed return values (not `def`)

File: grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy [hunk 6]
```diff
@@ -505,6 +505,10 @@ class PostProcessingService {
...
+        if (requestParams.disableAllQualityFilters) {
+            return [[:], [:], [:], [:]]
```
Comment (nickdos) [inline grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy:509]: Seems a big ugly to return 4 maps... would prefer a bean to be used as data transport class if data is this complex.

File: grails-app/services/au/org/ala/biocache/hubs/UserService.groovy [hunk 7]
```diff
@@ -40,4 +43,30 @@ class UserService {
             log.warn('could not find user details')
             return [displayName: userid, email: userid]
         }
-    }}
+    }
+
+    def getUserPref(String userId, HttpServletRequest request) {
```
Comment (nickdos) [inline grails-app/services/au/org/ala/biocache/hubs/UserService.groovy:48]: Ditto method return should be "typed"

File: grails-app/conf/plugin.groovy (hunk 2 above)
Comment (alexhuang091) [reply grails-app/conf/plugin.groovy:32]: I'll rever it then.

File: grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy (hunk 3 above)
Comment (alexhuang091) [reply grails-app/controllers/au/org/ala/biocache/hubs/OccurrenceController.groovy:67]: I'll check, basically anywhere dq related, check if dq is enabled.

File: grails-app/views/occurrence/list.gsp [hunk 8]
```diff
@@ -69,7 +69,11 @@
 ' Image does not support the identification of the species, subject is unclear and identifying features are difficult to see or not visible.<br/></div>',
             savePreferredSpeciesListUrl: "${createLink(controller: 'imageClient', action: 'saveImageToSpeciesList')}",
             getPreferredSpeciesListUrl:  "${createLink(controller: 'imageClient', action: 'getPreferredSpeciesImageList')}",
-            excludeCountUrl: "${createLink(controller: 'occurrence', action: 'dataQualityExcludeCounts', params: params.clone()).encodeAsJavaScript()}"
+            excludeCountUrl: "${createLink(controller: 'occurrence', action: 'dataQualityExcludeCounts', params: params.clone()).encodeAsJavaScript()}",
+            expandProfileDetails: ${expandProfileDetails},
+            userId: "${userId}",
+            prefKey: "${grailsApplication.config.dataquality.prefkey}",
```
Comment (nickdos) [inline grails-app/views/occurrence/list.gsp:75]: prefer `grailsApplication.config.getProperty("dataquality.prefkey", String, "")` to be used for these calls.

**Review by nickdos — COMMENTED** [review]
Looks good - will approve after comments have been addressed

File: grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy (hunk 6 above)
Comment (alexhuang091) [reply grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy:509]: Hi Nick, how about just a map.
so it's like

```
def result = [:]
if (disableAll) {
    return result
} 

result.val1 = val1
result.val2 = val2

return result
```

Comment (nickdos) [reply grails-app/services/au/org/ala/biocache/hubs/PostProcessingService.groovy:509]: Yeah, a plain Map is OK

**Review by nickdos — APPROVED** [review]

**Pull request merged**